core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -698,14 +695,24 @@ public Optional<FileRecords.TimestampAndOffset> findOffsetByTimestamp(TopicParti
                     && rlsMetadata.endOffset() >= startingOffset
                     && isRemoteSegmentWithinLeaderEpochs(rlsMetadata, unifiedLog.logEndOffset(), epochWithOffsets)
                     && rlsMetadata.state().equals(RemoteLogSegmentState.COPY_SEGMENT_FINISHED)) {
-                    return lookupTimestamp(rlsMetadata, timestamp, startingOffset);
+                    // cache to avoid race conditions
+                    List<LogSegment> segmentsCopy = new ArrayList<>(unifiedLog.logSegments());
+                    if (segmentsCopy.isEmpty() || rlsMetadata.startOffset() < segmentsCopy.get(0).baseOffset()) {
```

Comment:
So we're comparing the rlsMetadata start offset with current local log start offset here, right?

Reply:
yes, correct.